README.md
# Itamae::Plugin::Recipe::BedrockServer

Itamae plugin to install [Minecraft](https://www.minecraft.net/) with init scripts.
You must read [Minecraft Server Download](https://www.minecraft.net/en-us/download/server/bedrock/) in [Minecraft End User License Agreement](https://account.mojang.com/terms) and [Privacy Policy](https://privacy.microsoft.com/ja-jp/privacystatement).

[![Gem Version](https://badge.fury.io/rb/itamae-plugin-recipe-bedrock_server.svg)](https://badge.fury.io/rb/itamae-plugin-recipe-bedrock_server)
![test](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/workflows/test/badge.svg)
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](LICENSE)
![](https://ruby-gem-downloads-badge.herokuapp.com/itamae-plugin-recipe-bedrock_server?type=total)

## Supported
* Ubuntu 18 or later

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'itamae-plugin-recipe-bedrock_server'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install itamae-plugin-recipe-bedrock_server

## Usage

### Recipe

Install bedrock_server executable file and init scritps (systemd or init.d)

```ruby
# recipe.rb
include_recipe "bedrock_server"
```

Or `include_recipe` just what you need manually:

```ruby
include_recipe "bedrock_server::setup"
include_recipe "bedrock_server::install"
include_recipe "bedrock_server::service"
```

### Node

[default node.yml](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/blob/main/recipes/node.yml)

```yml
# node.yml
bedrock_server:
  # path to bedrock server directory (default: "/opt")
  app_dir: "/opt"

  # bedrock-server download url (defailt is check from https://minecraft.net/en-us/download/server/bedrock/ and detect latest url)
  url: "https://minecraft.azureedge.net/bin-linux/bedrock-server-1.16.201.02.zip"

  # bedrock_server service actions when after install (default: enable, start)
  service_actions:
    - enable
    - start
    # - disable
    # - stop

  # server.properties configuration
  configuration:
    server-name: "Dedicated Server"
    gamemode: "survival"
    difficulty: "easy"
    allow-cheats: false
    max-players: 10
    online-mode: true
    white-list: false
    server-port: 19132
    server-portv6: 19133
    view-distance: 32
    tick-distance: 4
    player-idle-timeout: 30
    max-threads: 8
    level-name: "Bedrock level"
    level-seed: ""
    level-type: "DEFAULT"
    default-player-permission-level: "member"
    texturepack-required: false
    content-log-file-enabled: false
    compression-threshold: 1
    server-authoritative-movement: "server-auth"
    player-movement-score-threshold: 20
    player-movement-distance-threshold: 0.3
    player-movement-duration-threshold-in-ms: 500
    correct-player-movement: false
```

## Development

```
    $ docker run --name tmp-ubuntu ubuntu:latest bash -c "apt-get update && apt-get install -y systemd-sysv"
    $ docker commit tmp-ubuntu ubuntu-with-systemd
    $ docker run --rm --privileged -d --name container-with-service -p 19132:19132 ubuntu-with-systemd /sbin/init
    $ bundle exec itamae docker --node-yaml=recipes/node.yml recipes/install.rb --container=container-with-service
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kyohah/itamae-plugin-recipe-bedrock_server. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Itamae::Plugin::Recipe::BedrockServer project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog

## [v0.2.0](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/tree/v0.2.0) (2021-01-23)

[Full Changelog](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/compare/v0.1.0...v0.2.0)

## [v0.1.0](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/tree/v0.1.0) (2021-01-22)

[Full Changelog](https://github.com/kyohah/itamae-plugin-recipe-bedrock_server/compare/116c84f7f2e3d15fc681a1fd3f9156d85949c8a3...v0.1.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


spec/spec_helper.rb
require 'serverspec'
require 'docker'
require "yaml"
require "itamae"

set :backend, :docker

set :docker_image, ENV['DOCKER_IMAGE']
set :docker_container, ENV['DOCKER_CONTAINER']

def node
  return @node if @node

  hash = YAML.load_file("#{__dir__}/../recipes/node.yml")

  @node = Itamae::Node.new(hash, Specinfra.backend)
end


spec/install_spec.rb
describe file("#{node[:bedrock_server][:app_dir]}/bedrock-server/bedrock_server") do
  it { should exist }
  it { should be_file }
  it { should be_executable }
end


spec/service_spec.rb
describe file("/etc/default/bedrock_server") do
  # TODO:
  let(:environment_file) do
    <<-EOS

LD_LIBRARY_PATH=#{node[:bedrock_server][:environment][:LD_LIBRARY_PATH]}
    EOS
  end

  it { should exist }
  it { should be_file }
  it { should be_mode "644" }
  its(:content) { should eq environment_file }
end

describe file("/etc/systemd/system/bedrock_server.service") do
  it { should exist }
  it { should be_file }
  it { should be_mode "644" }
  its(:content) { should include "ExecStart=#{node[:bedrock_server][:app_dir]}/bedrock-server/bedrock_server" }
  its(:content) { should include "WorkingDirectory=#{node[:bedrock_server][:app_dir]}/bedrock-server" }
end

describe file("/etc/bedrock_server.d") do
  it { should exist }
  it { should be_directory }
end

describe service("bedrock_server") do
  it { should be_enabled }
  it { should be_running }
end


lib/itamae/plugin/recipe/bedrock_server/setup.rb
hash = {
  'bedrock_server' => {
    'app_dir' => '/opt',
    'environment' => { 'LD_LIBRARY_PATH' => '.' },
    'service_actions' => %i[enable start],
    'configuration' => {
      'server-name' => 'Dedicated Server',
      'gamemode' => 'survival',
      'difficulty' => 'easy',
      'allow-cheats' => false,
      'max-players' => 10,
      'online-mode' => true,
      'white-list' => false,
      'server-port' => 19_132,
      'server-portv6' => 19_133,
      'view-distance' => 32,
      'tick-distance' => 4,
      'player-idle-timeout' => 30,
      'max-threads' => 8,
      'level-name' => 'Bedrock level',
      'level-seed' => '',
      'level-type' => 'DEFAULT',
      'default-player-permission-level' => 'member',
      'texturepack-required' => false,
      'content-log-file-enabled' => false,
      'compression-threshold' => 1,
      'server-authoritative-movement' => 'server-auth',
      'player-movement-score-threshold' => 20,
      'player-movement-distance-threshold' => 0.3,
      'player-movement-duration-threshold-in-ms' => 500,
      'correct-player-movement' => false
    }
  }
}

node.reverse_merge!(hash)

node.validate! do
  {
    bedrock_server: {
      configuration: {
        'server-name' => string,
        'gamemode' => match(/^(survival|creative|adventure)$/),
        'difficulty' => match(/^(peaceful|easy|normal|hard)$/),
        'allow-cheats' => boolean,
        'max-players' => integer,
        'online-mode' => boolean,
        'white-list' => boolean,
        'server-port' => integer,
        'server-portv6' => integer,
        'view-distance' => integer,
        'tick-distance' => integer,
        'player-idle-timeout' => integer,
        'max-threads' => integer,
        'level-name' => string,
        'level-seed' => optional(string),
        'level-type' => match(/^(FLAT|LEGACY|DEFAULT)$/),
        'default-player-permission-level' => match(/^(visitor|member|operator)$/),
        'texturepack-required' => boolean,
        'content-log-file-enabled' => boolean,
        'compression-threshold' => integer,
        'server-authoritative-movement' => match(/^(client-auth|server-auth)$/),
        'player-movement-score-threshold' => integer,
        'player-movement-distance-threshold' => float,
        'player-movement-duration-threshold-in-ms' => integer,
        'correct-player-movement' => boolean
      }
    }
  }
end
